backend/src/main/java/com/programmers/pcquotation/domain/member/service/OAuth2Service.kt

```diff
@@ -0,0 +1,317 @@
+import org.springframework.core.ParameterizedTypeReference
+import org.springframework.http.*
+import org.springframework.http.HttpMethod.POST
+import org.springframework.stereotype.Service
+import org.springframework.util.LinkedMultiValueMap
+import org.springframework.util.MultiValueMap
+import org.springframework.web.client.exchange
+
+
+@Slf4j
+@Service
+class OAuth2Service(
+    private val restTemplateConfig: RestTemplateConfig,
+    private val oAuthRegistration: OAuth2RegistrationProperties,
+    private val oAuthProvider: OAuth2ProviderProperties,
+    private val authService: AuthService,
+    private val customerService: CustomerService,
+) {
+
+
+    fun defaultHeader(): HttpHeaders {
+        val headers = HttpHeaders()
+        headers.contentType = MediaType.APPLICATION_FORM_URLENCODED;
+        headers.add("Content-type", "application/x-www-form-urlencoded;charset=utf-8");
+        return headers;
+    }
+
+    @Throws(RuntimeException::class)
+    fun kakaoLogin(authInfo: OAuth2AuthInfoDto) {
+        if (authInfo.code.isEmpty()) {
+            throw RedirectInfoException("Redirect information does not exist")
+        }
+        val kakaoTokenInfo = this.getKakaoTokenInfo(authInfo.code, authInfo.userType);
+        val userInfo = this.getKakaoUserInfo(kakaoTokenInfo.accessToken)
+        try {
+            when (authInfo.userType) {
+                CUSTOMER -> {
+                    val customer = customerService.findCustomerByEmail(userInfo.response.email)
+                    if (customer.isEmpty)
+                        authService.processCustomerSignup(userInfo.toCustomerSignupRequest())
+                    authService.processLoginCustomer(userInfo.toLoginRequest())
+                }
+                else -> {
+                    throw Oauth2LoginException("login_failed")
+                }
+
+            }
+        } catch (e: RuntimeException) {
+            throw Oauth2LoginException("login_failed")
+        }
+    }
+
+    private fun cvtObjectToMap(obj: Any): Map<String, Any> {
+        val mapper = ObjectMapper()
+        return mapper.convertValue<Map<String, Any>>(obj, object : TypeReference<Map<String, Any>>() {
+        })
+    }
+
+    @Throws(RuntimeException::class)
+    private fun getKakaoTokenInfo(authCode: String, userType: UserType): OAuth2TokenInfoDto {
+
+        var resultDto: OAuth2TokenInfoDto? = null
+        var responseTokenInfo: ResponseEntity<Map<String, Any>>? = null
+
+        // [STEP1] 카카오 토큰 URL로 전송할 데이터 구성
+        val requestParamMap: MultiValueMap<String, Any> = LinkedMultiValueMap()
+        requestParamMap.add("grant_type", "authorization_code");
+        requestParamMap.add("client_id", oAuthRegistration.kakao.clientId);
+        requestParamMap.add("redirect_uri", oAuthRegistration.kakao.redirectUri + "/" + userType.value.lowercase());
+        requestParamMap.add("code", authCode);
+        requestParamMap.add("client_secret", oAuthRegistration.kakao.clientSecret);
```

Comment:
```suggestion
        requestParamMap.apply {
                add("grant_type", "authorization_code")
                add("client_id", oAuthRegistration.kakao.clientId)
                add("redirect_uri", oAuthRegistration.kakao.redirectUri + "/" + userType.value.lowercase())
                add("code", authCode)
                add("client_secret", oAuthRegistration.kakao.clientSecret)
        }
```
이렇게 좀 더 보기 좋게 작성할 수도 있습니다!